inode/inode.go

```diff
@@ -351,13 +352,78 @@ func (vS *volumeStruct) inodeCacheTouch(inode *inMemoryInodeStruct) {
 	vS.Unlock()
 }
 
+// TODO: What happens if the volume is removed?  What if
+// failover?
+func (vS *volumeStruct) inodeCacheDiscard() {
+	inodesToDrop := uint64(0)
+
+	vS.Lock()
+
+	if (vS.inodeCacheLRUItems * globals.inodeSize) > vS.inodeCacheLRUMaxBytes {
+		// Check, at most, 1.25 * (minimum_number_to_drop)
+		inodesToDrop = (vS.inodeCacheLRUItems * globals.inodeSize) - vS.inodeCacheLRUMaxBytes
+		inodesToDrop = inodesToDrop / globals.inodeSize
+		inodesToDrop += inodesToDrop / 4
+	}
+	for ((vS.inodeCacheLRUItems * globals.inodeSize) > vS.inodeCacheLRUMaxBytes) &&
+		(inodesToDrop > 0) {
+		inodesToDrop--
+
+		ic := vS.inodeCacheLRUHead
+
+		// Create a DLM lock object
+		id := dlm.GenerateCallerID()
+		inodeRWLock, _ := InitInodeLock(vS.volumeName, ic.InodeNumber, id)
+		err := inodeRWLock.TryWriteLock()
+
+		// Inode is locked; skip it
+		if err != nil {
+			// Move inode to tail of LRU
+			vS.inodeCacheTouchWhileLocked(ic)
+			continue
+		}
+
+		if (ic.inFlightLogSegmentMap != nil) && (ic.dirty == false) {
+			// The inode is busy - drop the DLM lock and move to tail
+			inodeRWLock.Unlock()
+			vS.inodeCacheTouchWhileLocked(ic)
+			continue
+		}
+
+		vS.Unlock()
+
+		var ok bool
+
+		ok, err = vS.inodeCacheDrop(ic)
+		if err != nil || !ok {
+			pStr := fmt.Errorf("The inodes was not found in the inode cache - ok: %v err: %v", ok, err)
+			panic(pStr)
+		}
+
+		// TODO - I am releasing the locks out of order here.   That seems wrong.
+		// We acquire the locks in this order:
+		// 1. Volume lock
+		// 2. DLM lock for inode
```

Comment:
I think the reason it is ok to do (1) before (2) is because you are doing a TryLock for (2).
So I don't think it's actually wrong... Does this make sense?

Reply:
Makes sense - wanted to call it out so that we are all on the same page